File: effs/file_words_read_at.js
// Read up to max bytes at offset into a packed Array without moving the file.
function file_words_read_at(file, offset, max) {
  const fs = require("fs");
  const bytes = Buffer.alloc(max);
  try {
    const got = fs.readSync(file, bytes, 0, max, offset);
    const words = (max >>> 2) + 3;
    let cap = 1;
    while (cap < words) cap *= 2;
    const out = new Array(cap).fill(0);
    for (let i = 0; i < got; i += 4) {
      out[i >>> 2] = ((bytes[i] | 0) | ((bytes[i + 1] | 0) << 8) |
        ((bytes[i + 2] | 0) << 16) | ((bytes[i + 3] | 0) << 24)) >>> 0;
    }
    return io_tup(file, io_done(io_tup(out, got)));
  } catch (e) {
    return io_tup(file, io_fail(Math.abs(e.errno ?? 5)));
  }
}

File: effs/file_words_read_at.c
// Read up to max bytes at offset into a packed Array without moving the
// file position. The open File handle and the packed byte count return.
Term file_words_read_at_run(Env e, Term* f, IoWork* w) {
  int fd = (int)io_hand_v(f[0]);
  uint64_t off = (uint64_t)(u32)f[1];
  uint64_t max = (uint64_t)(u32)f[2];
  uint64_t words = (max >> 2) + 3;
  Nat d = 0;
  while ((1ull << d) < words) d += 1;

  Term arr = blk_new(e, 0, d, 0, 0, NULL);
  uint8_t* data = (uint8_t*)blk_ptr(e.mem, term_loc(arr), 0);
  uint64_t got = 0;
  while (got < max) {
    ssize_t n = pread(fd, data + got, max - got, (off_t)(off + got));
    if (n < 0 && errno == EINTR) continue;
    if (n < 0) {
      return io_tup(e, io_hand(fd), io_fail(e, (u32)errno, NULL));
    }
    if (n == 0) break;
    got += (uint64_t)n;
  }
  return io_tup(e, io_hand(fd), io_done(e, io_tup(e, arr, (Term)got)));
}

static void __attribute__((constructor)) file_words_read_at_use(void) {
  io_eff(CID_FILE_READ_WORDS_AT, file_words_read_at_run, 0);
}
